File: src/main/java/org/vanautrui/languages/compiler/parsing/astnodes/nonterminal/upperscopes/NamespaceNode.java
package org.vanautrui.languages.compiler.parsing.astnodes.nonterminal.upperscopes;

import org.vanautrui.languages.compiler.parsing.IASTNode;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public final class NamespaceNode implements IASTNode {

	//a namespace is represented by a filename.
	//the contents of a namespace are the contents of the file

	public Path srcPath;

	public String name;

	//structs must be declared before the subroutines
	public List<StructDeclNode> structs;

	public List<MethodNode> methods;

	public NamespaceNode(){}

	public NamespaceNode(final String name){
		this.name=name;
		this.srcPath = Paths.get("/dev/null");
		this.methods=new ArrayList<>();
		this.structs=new ArrayList<>();
	}

	public NamespaceNode(final MethodNode method, final String namespacename) {
		//utility method, to create a NamespaceNode from a single method
		this.name = namespacename;
		this.srcPath = Paths.get("/dev/null");
		this.methods=new ArrayList<>();
		this.structs=new ArrayList<>();
		this.methods.add(method);
	}


	@Override
	public String toSourceCode() {

		return Stream.concat(structs.stream(),methods.stream())
				.map(IASTNode::toSourceCode)
				.collect(Collectors.joining("\n"));

	}
}

File: dragon-parser/src/main/compiler/parsing/astnodes/nonterminal/upperscopes/NamespaceNode.cpp

//standard headers
#include <vector>
#include <set>

//project headers
#include "NamespaceNode.hpp"

public final class NamespaceNode implements IASTNode {

	//a namespace is represented by a filename.
	//the contents of a namespace are the contents of the file

	public final Path srcPath;

	public final String name;

	//structs must be declared before the subroutines
	public final List<StructDeclNode> structs = new ArrayList<>();

	public final List<MethodNode> methods = new ArrayList<>();

	public NamespaceNode(
			final TokenList tokens,
			final String namespace,
			final boolean debug
	) throws Exception {

		if (debug) {
			System.out.println("try to parse " + this.getClass().getSimpleName() + " from " + tokens.toSourceCodeFragment());
		}

		this.srcPath = Paths.get("/dev/null");
		this.name = namespace;
		final TokenList copy = tokens.copy();

		if(copy.size()>0) {

			IToken next_struct = copy.get(0);

			while (next_struct instanceof StructToken) {
				this.structs.add(new StructDeclNode(copy, debug));

				if (copy.size() > 0) {
					next_struct = copy.get(0);
				} else {
					break;
				}
			}
		}

		//it is be nice to have prefix 'fn' before a function
		//to make parsing easier.
		//this does not add much boilerplate to the syntax
		//and would probably make the parser faster
		if (copy.size() > 0) {

			IToken next_subr = copy.get(0);

			while (next_subr instanceof FnToken) {
				this.methods.add(new MethodNode(copy, debug));

				if (copy.size() > 0) {
					next_subr = copy.get(0);
				} else {
					break;
				}
			}
		}

		tokens.set(copy);
	}

}
